File: sample/py/11_lissajous_particles.py
# 11: リサージュ運動する粒子をGIF出力するサンプル。

from __future__ import annotations

import math
from time import perf_counter

from pytra.runtime.gif import save_gif


def color_palette() -> bytes:
    p = bytearray()
    for i in range(256):
        r = i
        g = (i * 3) % 256
        b = 255 - i
        p.append(r)
        p.append(g)
        p.append(b)
    return bytes(p)


def run_11_lissajous_particles() -> None:
    w = 320
    h = 240
    frames_n = 360
    particles = 48
    out_path = "sample/out/11_lissajous_particles.gif"

    start = perf_counter()
    frames: list[bytes] = []

    for t in range(frames_n):
        frame = bytearray(w * h)

        for p in range(particles):
            phase = p * 0.261799
            x = int((w * 0.5) + (w * 0.38) * math.sin(0.11 * t + phase * 2.0))
            y = int((h * 0.5) + (h * 0.38) * math.sin(0.17 * t + phase * 3.0))
            color = 30 + (p * 9) % 220

            for dy in range(-2, 3):
                for dx in range(-2, 3):
                    xx = x + dx
                    yy = y + dy
                    if xx >= 0 and xx < w and yy >= 0 and yy < h:
                        d2 = dx * dx + dy * dy
                        if d2 <= 4:
                            idx = yy * w + xx
                            v = color - d2 * 20
                            v = max(0, v)
                            if v > frame[idx]:
                                frame[idx] = v

        frames.append(bytes(frame))

    save_gif(out_path, w, h, frames, color_palette(), delay_cs=3, loop=0)
    elapsed = perf_counter() - start
    print("output:", out_path)
    print("frames:", frames_n)
    print("elapsed_sec:", elapsed)


if __name__ == "__main__":
    run_11_lissajous_particles()

File: sample/lua/11_lissajous_particles.lua
-- from __future__ import annotations as annotations (not yet mapped)
local math = math
-- from time import perf_counter as perf_counter (not yet mapped)
-- from pytra.runtime.gif import save_gif as save_gif (not yet mapped)

-- 11: Sample that outputs Lissajous-motion particles as a GIF.

function color_palette()
    p = bytearray()
    for i = 0, (256) - 1, 1 do
        r = i
        g = ((i * 3) % 256)
        b = (255 - i)
        p:append(r)
        p:append(g)
        p:append(b)
    end
    return bytes(p)
end

function run_11_lissajous_particles()
    w = 320
    h = 240
    frames_n = 360
    particles = 48
    out_path = "sample/out/11_lissajous_particles.gif"
    
    start = perf_counter()
    local frames = {  }
    
    for t = 0, (frames_n) - 1, 1 do
        frame = bytearray((w * h))
        local __hoisted_cast_1 = float(t)
        
        for p = 0, (particles) - 1, 1 do
            phase = (p * 0.261799)
            x = int(((w * 0.5) + ((w * 0.38) * math.sin(((0.11 * __hoisted_cast_1) + (phase * 2.0))))))
            y = int(((h * 0.5) + ((h * 0.38) * math.sin(((0.17 * __hoisted_cast_1) + (phase * 3.0))))))
            color = (30 + ((p * 9) % 220))
            
            for dy = (-2), (3) - 1, 1 do
                for dx = (-2), (3) - 1, 1 do
                    xx = (x + dx)
                    yy = (y + dy)
                    if ((xx >= 0) and (xx < w) and (yy >= 0) and (yy < h)) then
                        d2 = ((dx * dx) + (dy * dy))
                        if (d2 <= 4) then
                            idx = ((yy * w) + xx)
                            v = (color - (d2 * 20))
                            v = max(0, v)
                            if (v > frame[(idx) + 1]) then
                                frame[(idx) + 1] = v
                            end
                        end
                    end
                end
            end
        end
        frames:append(bytes(frame))
    end
    save_gif(out_path, w, h, frames, color_palette())
    elapsed = (perf_counter() - start)
    print("output:", out_path)
    print("frames:", frames_n)
    print("elapsed_sec:", elapsed)
end


run_11_lissajous_particles()
